src/main/java/org/folio/oaipmh/helpers/GetOaiRecordsHelper.java

```diff
@@ -0,0 +1,138 @@
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import me.escoffier.vertx.completablefuture.VertxCompletableFuture;
+import org.apache.log4j.Logger;
+import org.folio.oaipmh.Request;
+import org.folio.oaipmh.ResponseHelper;
+import org.folio.rest.tools.client.Response;
+import org.folio.rest.tools.client.interfaces.HttpClientInterface;
+import org.folio.rest.tools.utils.TenantTool;
+import org.openarchives.oai._2.ListRecordsType;
+import org.openarchives.oai._2.OAIPMH;
+import org.openarchives.oai._2.OAIPMHerrorType;
+import org.openarchives.oai._2.OAIPMHerrorcodeType;
+import org.openarchives.oai._2.RecordType;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static org.folio.rest.jaxrs.resource.Oai.GetOaiRecordsResponse.respond200WithApplicationXml;
+import static org.folio.rest.jaxrs.resource.Oai.GetOaiRecordsResponse.respond400WithApplicationXml;
+import static org.folio.rest.jaxrs.resource.Oai.GetOaiRecordsResponse.respond404WithApplicationXml;
+import static org.folio.rest.jaxrs.resource.Oai.GetOaiRecordsResponse.respond422WithApplicationXml;
+import static org.openarchives.oai._2.OAIPMHerrorcodeType.BAD_ARGUMENT;
+import static org.openarchives.oai._2.OAIPMHerrorcodeType.BAD_RESUMPTION_TOKEN;
+import static org.openarchives.oai._2.OAIPMHerrorcodeType.CANNOT_DISSEMINATE_FORMAT;
+import static org.openarchives.oai._2.VerbType.LIST_RECORDS;
+
+public class GetOaiRecordsHelper extends AbstractHelper {
+
+  private static final Logger logger = Logger.getLogger(GetOaiRecordsHelper.class);
+  private static final String GENERIC_ERROR = "Error happened while processing ListRecords verb request";
+
+  @Override
+  public CompletableFuture<javax.ws.rs.core.Response> handle(Request request, Context ctx) {
+    CompletableFuture<javax.ws.rs.core.Response> future = new VertxCompletableFuture<>(ctx);
+    try {
+      // 1. Validate request
+      List<OAIPMHerrorType> errors = validateListRequest(request);
+      if (!errors.isEmpty()) {
+        OAIPMH oai = buildOaiResponse(request).withErrors(errors);
+        future.complete(buildValidationFailureResponse(oai));
+        return future;
+      }
+
+      HttpClientInterface httpClient = getOkapiClient(request.getOkapiHeaders());
+
+      // 2. Search for instances
+      httpClient.request(storageHelper.buildItemsEndpoint(request), request.getOkapiHeaders(), false)
+        // 3. Verify response and build list of identifiers
+        .thenApply(response -> buildRecords(request, response, ctx))
+        .thenApply(records -> {
+          if (records == null) {
+            return buildNoRecordsFoundOaiResponse(request);
+          } else {
+            // 4. Now we need to get marc data for each instance and update record with metadata
+            return buildSuccessResponse(buildOaiResponse(request)
```

Comment:
Once the changes for [MODOAIPMH-17](https://issues.folio.org/browse/MODOAIPMH-17) are available, the logic here will be updated to make async calls to get metadata